### ISSUE TITLE
[REGRESSION Ember 2.0.2 -> Ember 2.1.0] Component attributes cannot be updated during initialization using this.set

### ISSUE THREAD

**Issue by OrKoN** [emberjs/ember.js#12556]
JSBin demonstrating the issue: http://emberjs.jsbin.com/yinuqe/edit?html,js,console,output
Related issue: https://github.com/emberjs/ember.js/issues/11559

```js
var MyComponent = Ember.Component.extend({
  test: null,
  init: function() {
    this._super();
    this.set('test', 'updated'); 
  }
});
```
`test` should be updated here, but it's not if component is called like shown below:

```hbs
{{my-component test="initial-value"}}
```

**Comment by quaertym** [comment]
Maybe related https://github.com/emberjs/ember.js/issues/12514

**Comment by rwjblue** [comment]
Yes, @quaertym I believe that they probably are the same root issue.

**Comment by rwjblue** [comment]
The issue is that the attrs proxy is copying the initial values from `this.attrs.foo` to `this.foo` just after `init` is called (but still from within the constructor).  I think I have a potential solution that will also mean a small work reduction for initial render.  Will try to update in the next day or so...

**Comment by rwjblue** [comment]
Looks like this functions properly on beta and canary branches.

**Comment by rwjblue** [comment]
I submitted https://github.com/emberjs/ember.js/pull/12559 to ensure we had a test that would fail if this fails again.

**Comment by rwjblue** [comment]
Working demos:

* beta: http://emberjs.jsbin.com/xefipebafu/edit?html,js,output
* canary: http://emberjs.jsbin.com/kiraki/edit?html,js,output

**Comment by OrKoN** [comment]
Thanks @rwjblue 

I have found another use case (very similar) that does not work even on canary (and 2.1.0 too, worked fine on 2.0.2): http://emberjs.jsbin.com/qodoli/edit?html,js,output 

Let my try to add a test myself.

**Comment by OrKoN** [comment]
https://github.com/emberjs/ember.js/issues/12556 sorry, I have submitted blindly because ember's dependencies are not compatible with my node version(v4.2.2) - could not run tests locally.

**Comment by runspired** [comment]
This bug was likely fixed in 2.2, I've bumped into it in Flexi where the following code fails to properly set `label` in 2.1, but correctly sets the label in all other versions from at least 1.13 to 2.6 alpha.

```
  init() {
    this._super();

    if (!this.label) {
      this.label = this.component;
    } else {
      this.label = `${this.component}:${this.label}`;
    }

    let properties = this.getProperties('label', 'component', 'model', 'copy', 'expires');

    // properties.label is correct
    this.get('sustains').install(properties);
  },

.... some time later ...

willInsertElement() {
  this.get('label');  // incorrect value
}


```

**Comment by locks** [comment]
I am closing this issue as the above message mentions the particular issue should be fixed now.
Feel free to re-open the issue if you think it's still justified, thank you!

**Comment by runspired** [comment]
@locks I've previously confirmed that the fix should be back ported to 2.1

**Comment by locks** [comment]
Oops. I'm still leaving it closed, as 2.4 LTS has been released and that's currently how far back things are likely to be backported.
